types/chainLock.go

```diff
@@ -0,0 +1,123 @@
+package types
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/tendermint/tendermint/crypto"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+type ChainLock struct {
+	CoreBlockHeight    uint32          `json:"core_block_height,string,omitempty"`   // height of Chain Lock.
+	CoreBlockHash      []byte         `json:"core_block_hash,string,omitempty"`     // hash of Chain Lock.
+	Signature          []byte         `json:"signature,string,omitempty"`           // signature.
+}
+
+// ToProto converts Header to protobuf
+func (cl *ChainLock) ToProto() *tmproto.ChainLock {
+	if cl == nil {
+		return nil
+	}
+
+	return &tmproto.ChainLock{
+		CoreBlockHeight:         cl.CoreBlockHeight,
+		CoreBlockHash:           cl.CoreBlockHash,
+		Signature:               cl.Signature,
+	}
+}
+
+func (cl *ChainLock) PopulateFromChainLockParams(clp tmproto.ChainLockParams) error {
+
+	cl.CoreBlockHeight = clp.NewestChainLock.ChainLockHeight
+	cl.CoreBlockHash = clp.NewestChainLock.ChainLockHash
+	cl.Signature = clp.NewestChainLock.Signature
+
+	return cl.ValidateBasic()
+}
+
+func (cl ChainLock) RequestId() []byte {
+	s := []byte{0x05, 0x63, 0x6c, 0x73, 0x69, 0x67} //5 clsig
+
+	var coreBlockHeightBytes [4]byte
+	binary.LittleEndian.PutUint32(coreBlockHeightBytes[:],cl.CoreBlockHeight)
+
+
+	s = append(s,coreBlockHeightBytes[:]...)
+	return crypto.Sha256(crypto.Sha256(s))
+}
+
+// ValidateBasic performs stateless validation on a Chain Lock returning an error
+// if any validation fails.
+// It does not verify the signature
+func (cl ChainLock) ValidateBasic() error {
+
+	if cl.CoreBlockHeight == 0 {
+		return errors.New("zero Chain Lock Height")
+	}
+
+	if len(cl.CoreBlockHash) == 0 {
+		return fmt.Errorf("chain lock block hash is empty")
+	}
+
+	if err := ValidateHash(cl.CoreBlockHash); err != nil {
+		return fmt.Errorf("chain lock block hash is wrong size")
+	}
+
+	if len(cl.Signature) == 0 {
+		return fmt.Errorf("chain lock signature is empty")
+	}
+
+	if err := ValidateSignatureSize(crypto.BLS12381, cl.Signature); err != nil {
+		return fmt.Errorf("chain lock signature is wrong size")
+	}
+
+	return nil
+}
+
+// StringIndented returns a string representation of the header
+func (cl *ChainLock) StringIndented(indent string) string {
+	if cl == nil {
+		return "nil-ChainLock"
+	}
+	return fmt.Sprintf(`ChainLock{
+%s  CoreBlockHeight:   %v
+%s  CoreBlockHash:     %v
+%s  Signature:         %v}`,
+		indent, cl.CoreBlockHeight,
+		indent, cl.CoreBlockHash,
+		indent, cl.Signature )
+}
+
+// FromProto sets a protobuf Header to the given pointer.
+// It returns an error if the chain lock is invalid.
+func ChainLockFromProto(clp *tmproto.ChainLock) (*ChainLock, error) {
+	if clp == nil {
+		return nil, nil
+	}
+
+	cl := new(ChainLock)
+	cl.CoreBlockHeight = clp.CoreBlockHeight
+	cl.CoreBlockHash = clp.CoreBlockHash
+	cl.Signature = clp.Signature
+
+	return cl, cl.ValidateBasic()
+}
+
+func NewMockChainLock() ChainLock {
```

Comment:
NewMockCoreChainLock

Reply:
It's a chain lock, we only have 1 type in Dash